src/script/builder/tir/stmt.cc

```diff
@@ -0,0 +1,265 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "./stmt.h"
+
+#include <tvm/runtime/registry.h>
+#include <tvm/tir/op.h>
+
+#include "./prim_func_frame.h"
+#include "./var.h"
+
+namespace tvm {
+namespace script {
+namespace builder {
+namespace tir {
+
+void AssertFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::AssertStmt(condition, message, AsStmt(stmts)));
+}
+
+void LetFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::LetStmt(var, value, AsStmt(stmts)));
+}
+
+void AllocateFrameNode::ExitWithScope() {
+  using namespace tvm::tir;
+  TIRFrameNode::ExitWithScope();
+  Buffer flattened_buffer = buffer.GetFlattenedBuffer();
+  AddToParent(Allocate(buffer->data, flattened_buffer->dtype, flattened_buffer->shape, condition,
+                       AsStmt(stmts), annotations));
+}
+
+void AllocateConstFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::AllocateConst(buffer->data, dtype, extents, data, AsStmt(stmts)));
+}
+
+void LaunchThreadFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::AttrStmt(env_var, attr_key, extent, AsStmt(stmts)));
+}
+
+void RealizeFrameNode::ExitWithScope() {
+  using namespace tvm::tir;
+  TIRFrameNode::ExitWithScope();
+  AddToParent(AttrStmt(
+      buffer_slice->buffer, "realize_scope", StringImm(storage_scope_str),
+      BufferRealize(buffer_slice->buffer, buffer_slice->region, condition, AsStmt(stmts))));
+}
+
+void AttrFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::AttrStmt(node, attr_key, value, AsStmt(stmts)));
+}
+
+void WhileFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  AddToParent(tvm::tir::While(condition, AsStmt(stmts)));
+}
+
+void IfFrameNode::ExitWithScope() {
+  TIRFrameNode::ExitWithScope();
+  if (!stmts.empty()) {
+    LOG(FATAL) << "stmt within IfThenElse frame should be either in ThenFrame or ElseFrame";
+  }
+  if (then_stmt.empty()) {
+    LOG(FATAL) << "IfThenElse frame should have at least one then branch";
+  }
+  AddToParent(
+      tvm::tir::IfThenElse(condition, AsStmt(then_stmt),
+                           else_stmt.empty() ? tvm::tir::Stmt(nullptr) : AsStmt(else_stmt)));
+}
+
+void ThenFrameNode::ExitWithScope() {
```

Comment:
Let's also check in `ThenFrameNode::EnterWithScope`, and error out if `then_stmts` is already defined.

Example of overloading `EnterWithScope`: https://github.com/junrushao1994/tvm/blob/054ef7e78456234a5942c66e63acda0249af1c7f/src/script/builder/tir/block_frame.cc#L69-L73. Just don't forget to call `parent-class::EnterWithScope()`